promql/analyzer.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2013 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package promql
+
+import (
+	"errors"
+	"time"
+
+	"golang.org/x/net/context"
+
+	clientmodel "github.com/prometheus/client_golang/model"
+
+	"github.com/prometheus/prometheus/storage/local"
+)
+
+// An Analyzer traverses an expression and determines which data has to be requested
+// from the storage. It is bound to a context that allows cancellation and timing out.
+type Analyzer struct {
+	ng *Engine // The engine that launched the Analyzer.
+
+	Expr       Expr                  // The expression being analyzed.
+	Start, End clientmodel.Timestamp // The time range for evaluation of Expr.
+
+	offsetPreloadTimes map[time.Duration]preloadTimes
+}
+
+// preloadTimes tracks which instants or ranges to preload for a set of
+// fingerprints. One of these structs is collected for each offset by the query
+// analyzer.
+type preloadTimes struct {
+	// Instants require single samples to be loaded along the entire query
+	// range, with intervals between the samples corresponding to the query
+	// resolution.
+	instants map[clientmodel.Fingerprint]struct{}
+	// Ranges require loading a range of samples at each resolution step,
+	// stretching backwards from the current evaluation timestamp. The length of
+	// the range into the past is given by the duration, as in "foo[5m]".
+	ranges map[clientmodel.Fingerprint]time.Duration
+}
+
+// analyze the provided expression and attach metrics and fingerprints to data-selecting
+// AST nodes that are later used to preload the data from the storage.
+func (a *Analyzer) analyze(ctx context.Context) {
+	a.offsetPreloadTimes = map[time.Duration]preloadTimes{}
+
+	getPreloadTimes := func(offset time.Duration) preloadTimes {
+		if _, ok := a.offsetPreloadTimes[offset]; !ok {
+			a.offsetPreloadTimes[offset] = preloadTimes{
+				instants: map[clientmodel.Fingerprint]struct{}{},
+				ranges:   map[clientmodel.Fingerprint]time.Duration{},
+			}
+		}
+		return a.offsetPreloadTimes[offset]
+	}
+
+	// Retrieve fingerprints and metrics for the required time range for
+	// each metric or matrix selector node.
+	Inspect(a.Expr, func(node Node) bool {
+		switch n := node.(type) {
+		case *VectorSelector:
+			pt := getPreloadTimes(n.Offset)
+			fpts := a.ng.storage.GetFingerprintsForLabelMatchers(n.LabelMatchers)
```

Comment:
Hmm, reaching multiple levels into an object's members is something we've tried to avoid in the past, as it's a [http://en.wikipedia.org/wiki/Law_of_Demeter](Law of Demeter) violation. Seems the only reason why you don't hand in the storage directly anymore (but the engine) is that you also need the `checkContext()` method from the analyzer. Is there a way to reorg that to not need the multi-level "reach-ins"? Like passing in `checkContext` seperately as a callback to the analyzer?